File: helpers/python/debugProbe.py
import datetime
import json
import os
import urllib.error
import urllib.request


def debug_probe(label, data=None, meta=None):
    try:
        if os.getenv("DEBUGHUB_ENABLED") != "1":
            return

        session_id = os.getenv("DEBUGHUB_SESSION")
        if not session_id:
            return

        endpoint = os.getenv("DEBUGHUB_ENDPOINT")
        if not endpoint:
            return

        meta = meta or {}
        level = meta.get("level") or "info"
        if level not in ("info", "warn", "error"):
            level = "info"

        event = {
            "ts": datetime.datetime.now(datetime.timezone.utc).isoformat().replace("+00:00", "Z"),
            "sessionId": session_id,
            "label": label,
            "data": data if data is not None else None,
            "hypothesisId": meta.get("hypothesisId"),
            "loc": meta.get("loc"),
            "level": level,
            "tags": meta.get("tags") if isinstance(meta.get("tags"), dict) else None,
            "runtime": "python",
        }

        payload = json.dumps(event).encode("utf-8")
        target = endpoint.rstrip("/") + "/event"

        req = urllib.request.Request(
            target,
            data=payload,
            headers={"Content-Type": "application/json"},
            method="POST",
        )
        with urllib.request.urlopen(req, timeout=2):
            pass
    except Exception:
        return


def debugProbe(label, data=None, meta=None):
    debug_probe(label, data, meta)

File: helpers/php/DebugProbe.php
<?php

function debug_probe($label, $data = null, $meta = null) {
    try {
        if (getenv('DEBUGHUB_ENABLED') !== '1') {
            return;
        }

        $sessionId = getenv('DEBUGHUB_SESSION');
        if (!$sessionId) {
            return;
        }

        $endpoint = getenv('DEBUGHUB_ENDPOINT');
        if (!$endpoint) {
            return;
        }

        if (!is_array($meta)) {
            $meta = [];
        }

        $level = isset($meta['level']) ? $meta['level'] : 'info';
        if (!in_array($level, ['info', 'warn', 'error'], true)) {
            $level = 'info';
        }

        $tags = null;
        if (isset($meta['tags']) && is_array($meta['tags'])) {
            $tags = $meta['tags'];
        }

        $event = [
            'ts' => gmdate('Y-m-d\TH:i:s\Z'),
            'sessionId' => $sessionId,
            'label' => $label,
            'data' => $data,
            'hypothesisId' => isset($meta['hypothesisId']) ? $meta['hypothesisId'] : null,
            'loc' => isset($meta['loc']) ? $meta['loc'] : null,
            'level' => $level,
            'tags' => $tags,
            'runtime' => 'php',
        ];

        $payload = json_encode($event, JSON_UNESCAPED_SLASHES);
        if ($payload === false) {
            return;
        }

        $url = rtrim($endpoint, '/') . '/event';
        $context = stream_context_create([
            'http' => [
                'method' => 'POST',
                'header' => "Content-Type: application/json\r\n",
                'content' => $payload,
                'timeout' => 2,
            ],
        ]);

        @file_get_contents($url, false, $context);
    } catch (Throwable $err) {
        return;
    }
}

function debugProbe($label, $data = null, $meta = null) {
    debug_probe($label, $data, $meta);
}
